### PR TITLE
Fix gfortran and e-antic builds with XCode 16 on x86_64

### PR DESCRIPTION
**Pull request by culler**
This PR modifies the patches in the gfortran spkg and adds a patch to the boost_cropped spkg. It addresses some issues caused by changes added in XCode 16, which is the XCode version compatible with macOS 15 Sequoia.

The gfortran spkg failed to build on macOS 15 Sequoia with XCode 16 because of [this bug](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=116809) which is specific to clang 16.  The bug report contains a patch, which is what this PR adds.  However, that patch changed files which were also being patched by the existing gcc-13.3.0-arm.patch.  Rather than have patches which must be applied in a certain order, this PR replaces the gcc-13.3.0-arm patch with a new patch file named gcc-13_3_0.patch which incorporates the new changes (no longer just for arm) with the older ones.

The e-antic spkg failed to build on macOS 15 Sequoia.  The reason turned out to be that XCode 16, unlike earlier versions, now generates an error for code which used to generate a warning about assigning an out-of-range value to an enum.  The code that triggers the error is in a boost header file.  There is a patch for the header file in [this macports ticket](https://trac.macports.org/ticket/68920).  The patch modifies a line in the header file which controls how this warning / error is handled by certain compilers.  The change adds clang 16 to the list of those compilers.   

See [https://groups.google.com/g/sage-devel/c/sv3jCczVa40](https://groups.google.com/g/sage-devel/c/sv3jCczVa40l) and [https://groups.google.com/g/sage-devel/c/yYRBm-hc918](https://groups.google.com/g/sage-devel/c/yYRBm-hc918)

### :memo: Checklist

- [x] The title is concise and informative.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation and checked the documentation preview.

### CODE REVIEW

**Comment by jhpalmieri** [comment]
I just built `e_antic` without this PR, so maybe there are other differences contributing to success or failure of the build. This machine: Sequoia 15.0.1, Xcode 16.0.

I'll try `gfortran` next.

**Comment by jhpalmieri** [comment]
Oh, maybe the issue is that I'm using a `homebrew` version of `boost`. Let me try again with Sage's `boost_cropped`.

**Comment by jhpalmieri** [comment]
> Oh, maybe the issue is that I'm using a `homebrew` version of `boost`. Let me try again with Sage's `boost_cropped`.

Yes, if I do `./configure --enable-e_antic --with-system-boost-cropped=no`, then `e_antic` only builds when using this PR.

**Comment by culler** [comment]
Thank you, @jhpalmieri!  If it is easy to check, I would be curious to know whether homebrew modified the file boost/mpl/aux_/integral_wrapper.hpp.  (Does the homebrew version of that header contain the string `__clang_major__ >= 16`?)

**Comment by jhpalmieri** [comment]
> Thank you, @jhpalmieri! If it is easy to check, I would be curious to know whether homebrew modified the file boost/mpl/aux_/integral_wrapper.hpp. (Does the homebrew version of that header contain the string `__clang_major__ >= 16`?)

The string "clang" is not found in that file, nor in any file in that directory. Also, for what it's worth, no files in that directory have been changed (according to their modification times) since August.

**Comment by culler** [comment]
Thanks.  I'm not sure what it means, but it is interesting.

**Comment by culler** [comment]
Comparing the the gcc command which failed for you with the corresponding command from my log file (both shown below), the difference that I see is that my system is aarch64-apple-darwin24.1.0 while yours is aarch64-apple-darwin24.0.0.  The reason for that is that I updated my Sequoia beta release before trying to install the official Sequoia release.  The Apple installer then refused to let me do the "downgrade" from macOS 15.1beta to macOS 15.0.

Maybe the internet knows something about the error you are seeing ...



`[spkg-build] /private/var/tmp/sage-10.5-current/local/var/tmp/sage/build/gfortran-13.3.0/gcc-build/./gcc/xgcc -B/private/var/tmp/sage-10.5-current/local/var/tmp/sage/build/gfortran-13.3.0/gcc-build/./gcc/ -B/private/var/tmp/sage-10.5-current/local/aarch64-apple-darwin24.1.0/bin/ -B/private/var/tmp/sage-10.5-current/local/aarch64-apple-darwin24.1.0/lib/ -isystem /private/var/tmp/sage-10.5-current/local/aarch64-apple-darwin24.1.0/include -isystem /private/var/tmp/sage-10.5-current/local/aarch64-apple-darwin24.1.0/sys-include    -g -O2 -mmacosx-version-min=11.0 -O2  -g -O2 -mmacosx-version-min=11.0 -DIN_GCC   -W -Wall -Wno-narrowing -Wwrite-strings -Wcast-qual -Wstrict-prototypes -Wmissing-prototypes -Wold-style-definition  -isystem ./include   -mmacosx-version-min=11 -mmacosx-version-min=11.0 -fno-common -g -DIN_LIBGCC2 -fbuilding-libgcc -fno-stack-protector   -mmacosx-version-min=11 -mmacosx-version-min=11.0 -fno-common -I. -I. -I../.././gcc -I../../../src/libgcc -I../../../src/libgcc/. -I../../../src/libgcc/../gcc -I../../../src/libgcc/../include -I../../../src/libgcc/config/libbid -DENABLE_DECIMAL_BID_FORMAT -DHAVE_CC_TLS -DUSE_EMUTLS  -o cas_16_1.o -MT cas_16_1.o -MD -MP -MF cas_16_1.dep -DL_cas -DSIZE=16 -DMODEL=1 -c ../../../src/libgcc/config/aarch64/lse.S`

`[spkg-build] /Users/palmieri/Sage/TESTING/sage-10.5.beta6/local/var/tmp/sage/build/gfortran-13.3.0/gcc-build/./gcc/xgcc -B/Users/palmieri/Sage/TESTING/sage-10.5.beta6/local/var/tmp/sage/build/gfortran-13.3.0/gcc-build/./gcc/ -B/Users/palmieri/Sage/TESTING/sage-10.5.beta6/local/aarch64-apple-darwin24.0.0/bin/ -B/Users/palmieri/Sage/TESTING/sage-10.5.beta6/local/aarch64-apple-darwin24.0.0/lib/ -isystem /Users/palmieri/Sage/TESTING/sage-10.5.beta6/local/aarch64-apple-darwin24.0.0/include -isystem /Users/palmieri/Sage/TESTING/sage-10.5.beta6/local/aarch64-apple-darwin24.0.0/sys-include    -g -O2 -O2  -g -O2 -DIN_GCC   -W -Wall -Wno-narrowing -Wwrite-strings -Wcast-qual -Wstrict-prototypes -Wmissing-prototypes -Wold-style-definition  -isystem ./include   -mmacosx-version-min=11 -mmacosx-version-min=11.0 -fno-common -g -DIN_LIBGCC2 -fbuilding-libgcc -fno-stack-protector   -mmacosx-version-min=11 -mmacosx-version-min=11.0 -fno-common -I. -I. -I../.././gcc -I../../../src/libgcc -I../../../src/libgcc/. -I../../../src/libgcc/../gcc -I../../../src/libgcc/../include -I../../../src/libgcc/config/libbid -DENABLE_DECIMAL_BID_FORMAT -DHAVE_CC_TLS -DUSE_EMUTLS  -o cas_16_1.o -MT cas_16_1.o -MD -MP -MF cas_16_1.dep -DL_cas -DSIZE=16 -DMODEL=1 -c ../../../src/libgcc/config/aarch64/lse.S`

**Comment by culler** [comment]
This looks like what you are seeing: [https://trac.macports.org/ticket/70710](https://trac.macports.org/ticket/70710).  They claim that it is a clang bug that has been fixed by llvm and will appear in XCode 17, but evidently it is fixed in my XCode 16.  They say:

"it's actually a bug in clang that was rolled back in upstream clang - you might want to try again with the Xcode RC, it probably has the fix. That said, gcc14 does build fine as long as configurepipe is set to no."

I don't know how to set configurepipe to no.

**Comment by culler** [comment]
It turns out that my macOS 15.1beta system has a beta version of XCode 16.1.  So I guess it was lucky that I was not allowed to "downgrade" macOS 15.1beta to macOS 15.0.

**Comment by jhpalmieri** [comment]
With the important caveat that I don't know what I'm doing, I managed to get Sage's `gfortran` to build. I took the patch from https://github.com/iains/gcc-darwin-arm64/issues/137#issuecomment-2386391940, which did not apply cleanly, modified it so it did apply cleanly, and then made a similar change in your new patch. The version of the other patch which does apply cleanly for me:
```diff
diff -Naur gcc-13.3-darwin-r0/libgcc/config/aarch64/lse.S gcc-13.3-darwin-r0-fix-cfi-startproc-error/libgcc/config/aarch64/lse.S
--- a/libgcc/config/aarch64/lse.S	2024-05-22 09:07:04
+++ b/libgcc/config/aarch64/lse.S	2024-09-30 18:06:18
@@ -170,8 +170,8 @@
 	.globl	\name
 	.hidden	\name
 	.type	\name, %function
-	.cfi_startproc
 \name:
+	.cfi_startproc
 	BTI_C
 .endm
 ```
The change in `gcc-13_3_0.patch`: move `.cfi_startproc` from before `_\name` to after it, so that part of the patch file looks like:
```diff
@@ -187,6 +192,29 @@ see the files COPYING3 and COPYING.RUNTIME respectively.  If not, see
 	cbz	w(tmp0), \label
 .endm
 
+#else
+.macro	STARTFN name
+	.text
+	.balign	16
+	.private_extern	_\name
+_\name:
+	.cfi_startproc
+	BTI_C
+.endm
+
+.macro	ENDFN name
+	.cfi_endproc
+.endm
+
+/* Branch to LABEL if LSE is disabled.  */
+.macro	JUMP_IF_NOT_LSE label
+	adrp	x(tmp0), ___aarch64_have_lse_atomics@PAGE
+	ldrb	w(tmp0), [x(tmp0), ___aarch64_have_lse_atomics@PAGEOFF]
+	cbz	w(tmp0), \label
+.endm
+
+#endif
+
 #ifdef L_cas
 
 STARTFN	NAME(cas)
```

**Comment by culler** [comment]
Wow.  That looks great.  I will test with XCode 16.1beta.

**Comment by dimpase** [comment]
On an Intel macOS 15.0.1 machine with an up do date Homebrew gfortran just works for me.
So all this (yet again) looks like a not very productive kind of PR, adding patches of 1600 lines, while the tool is there there to use.

**Comment by culler** [comment]
Your "1600 lines" is a completely meaningless, made-up number.  There are less than 10 lines of code that actually needed to change from the patched code in the previous version of this spkg to the patched code in this one.  One fairly large patch file was replaced by another patch file of about the same size, but with a different name.  So what?

This PR aims to make the gfortran and e-antic spkgs work with the latest XCode.  You have already made it clear that you are on a campaign to get rid of the gfortran spkg altogether. But that is irrelevant to this PR, and raising that old issue here is not at all helpful.  If you have nothing constructive to say, please go find someone else to harass.

I will say that your comment does add some helpful context to your claim that "PRs are welcome."  Yeah, right.  Visitors are always "welcome" when the attack dogs are hungry.

**Comment by dimpase** [comment]
my comment says that I cannot reproduce the claimed problem,  that gfortran is broken in the configuration you have. That's all.

**Comment by dimpase** [comment]
Useful PRs are welcome. Meaningless PRs are not. And no, I am not harassing you or anyone.
Why can't you just use Homebrew?

We should be doing useful things, e.g. adding Windows support to our cysignals.

**Comment by dimpase** [comment]
But I noticed that you casually called me "attack dog", @culler. Thanks, much appreciated.

**Comment by culler** [comment]
> my comment says that I cannot reproduce the claimed problem, that gfortran is broken in the configuration you have. That's all.

You don't have the same configuration.  I am not using homebrew.  Your comment suggests that your homebrew gfortran works fine.  This is about compiling gfortran with Sage's gfortran spkg.  If that is what you meant then your message was unclear.

**Comment by nbruin** [comment]
A reminder that this pull request is part of SageMath and therefore participants are required to adhere to the [SageMath Code of Conduct](https://github.com/sagemath/sage/blob/develop/CODE_OF_CONDUCT.md). 
Nils (on behalf of the code of conduct committee)

**Comment by culler** [comment]
> Useful PRs are welcome. Meaningless PRs are not. And no, I am not harassing you or anyone.

Yes, you are indeed harrassing me.

> Why can't you just use Homebrew?

That question has nothing to do with this PR, which is about Sage's e-antic and gfortran spkgs.

Also, it is not your business what I use.

**Comment by culler** [comment]
I refuse to be treated like this.  I am closing the PR.

**Comment by dimpase** [comment]
> Your "1600 lines" is a completely meaningless, made-up number.

This pacth, https://patch-diff.githubusercontent.com/raw/sagemath/sage/pull/38783.diff, has over 3K lines. 1600 is roughly how many lines of code it's changing. Are you even sure there is no backdoor brought in by it?

**Comment by dimpase** [comment]
In case one really needs an up to date gfortran on macOS without Homebrew, one can use https://github.com/fxcoudert/gfortran-for-macOS - which is more or less what Homebrew is using, and maintained by @fxcoudert in Homebrew; provided an a regular Apple way (a signed dmg installer).

Adding big semi-random patches of unknown provenance to older versions of gcc seems to be rather error-prone and a potential security hazard.
